### PR TITLE
Make controls react to changes in brushes.

### PR DESCRIPTION
**Pull request by grokys**
Previously when a property on a brush was changed (e.g. `SolidColorBrush.Color`), the control that used the brush didn't invalidate itself. This PR makes that happen. This is the first step towards being able to animate brushes.

There are a few steps to this:

- Added a `IMutableBrush.Changed` event that is triggered when a property that affects the rendering of the brush changes
- Added a `Brush.AffectsRender` property similar to that on `Visual` which raises the `Changed` event
- Make `GradientBrush.GradientStops` an `AvaloniaList` so changes to it can be monitored and raise the `Changed` event
- Added a `BrushAffectsRender` method to listen for changes on an `IBrush` property of a control and invalidate the control

Note that only controls that do their own rendering need to call `BrushAffectsRender` - templated controls have no visual representation of their own so they don't need to call it.

### CODE REVIEW

File: src/Avalonia.Visuals/Media/Brush.cs [hunk 1]
```diff
@@ -50,5 +53,36 @@ public static IBrush Parse(string s)
...
+        /// <summary>
+        /// Marks a property as affecting the brush's visual representation.
+        /// </summary>
+        /// <param name="properties">The properties.</param>
+        /// <remarks>
+        /// After a call to this method in a brush's static constructor, any change to the
+        /// property will cause the <see cref="Changed"/> event to be raised on the brush.
+        /// </remarks>
+        protected static void AffectsRender<T>(params AvaloniaProperty[] properties)
+            where T : Brush
```
Comment (aelij) [inline src/Avalonia.Visuals/Media/Brush.cs]: You don't need this type parameter. Just use `as Brush` below.

File: src/Avalonia.Visuals/Media/Immutable/ImmutableLinearGradientBrush.cs [hunk 2]
```diff
@@ -31,7 +31,7 @@ public class ImmutableLinearGradientBrush : ImmutableGradientBrush, ILinearGradi
         /// Initializes a new instance of the <see cref="ImmutableLinearGradientBrush"/> class.
         /// </summary>
         /// <param name="source">The brush from which this brush's properties should be copied.</param>
-        public ImmutableLinearGradientBrush(ILinearGradientBrush source)
```
Comment (aelij) [inline src/Avalonia.Visuals/Media/Immutable/ImmutableLinearGradientBrush.cs]: Why is this change required?

File: src/Avalonia.Visuals/Visual.cs [hunk 3]
```diff
@@ -328,6 +328,44 @@ protected static void AffectsRender(params AvaloniaProperty[] properties)
...
+        /// called.
+        /// </summary>
+        /// <param name="properties">The properties.</param>
+        /// <remarks>
+        /// This method should be called in a control's static constructor with each property
+        /// on the control which when changed should cause a redraw. It not only triggers an
+        /// invalidation when the property itself changes, but also when the brush raises
+        /// the <see cref="IMutableBrush.Changed"/> event.
+        /// </remarks>
+        protected static void BrushAffectsRender<T>(params AvaloniaProperty<IBrush>[] properties)
```
Comment (aelij) [inline src/Avalonia.Visuals/Visual.cs]: I think it might be better to introduce an interface like `IAffectsRender` (with a Changed event) that when implemented by a DP, the `AffectsRender` method automatically hooks into that event. That way the API remains agnostic to the fact that it's a brush. `Geometry` could also implement it, for example.

Comment (aelij) [reply src/Avalonia.Visuals/Visual.cs]: The `property.Changed` event is untyped anyhow. So why not just do the same thing you did here in the `AffectsRender` method? Only use a generic interface instead of `IMutableBrush`. For example:

```csharp
if (e.OldValue is IAffectsRender oldValue)
{
    oldValue.Changed -= ...
}
```

The only downside is that each value change for a property marked as AffectsRender would have two additional casts during assignment.

But the upside is that the API surface is smaller and less error-prone.

Comment (grokys) [reply src/Avalonia.Visuals/Visual.cs]: Ah, yes I missed the fact that you wanted to do it automatically in `AffectsRender`. We don't want to do that. For example, `Button` has a `Background` property but when that changes we don't want to call `InvalidateVisual` because `Button` doesn't actually render anything: the actual rendering is done by the `Border` or whatever in the `Button`'s template. I mentioned this in the PR body:

> Note that only controls that do their own rendering need to call BrushAffectsRender - templated controls have no visual representation of their own so they don't need to call it.

So I think it this behaviour needs to be opted-in.

This will become even more vital once we extend this pattern to `Geometry` as changing that can cause a remeasure and we don't want to queue more remeasures than necessary.

Comment (aelij) [reply src/Avalonia.Visuals/Visual.cs]: Yeah, but `Button` wouldn't have `AffectsRender` for background, right? Haven't you changed every `AffectsRender` that had a brush with it to `BrushAffectsRender`? If a property does affect the render, then also changes to its internal structure should. Same goes for geometries.

Comment (grokys) [reply src/Avalonia.Visuals/Visual.cs]: Ah yes, you're right of course. I seem to have had a brainfart ;) Sorry, that took me way to long to understand what you were saying.

Comment (grokys) [reply src/Avalonia.Visuals/Visual.cs]: Ok, make that change!

**Review by aelij — APPROVED** [review]
